Захаров Андрей

Folleach opened the pull request:
@goodoldlameme

Folleach committed:
Create CircularCloudLayouter and tests for it

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+            double angle,
+            Rectangle currentRectangle);
+
+        public event CircularLayoutState DebugStateChange;
```
не нашла, чтобы какой-то метод подписывался на это событие

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+
+        private bool CanPlaced(Rectangle targetRectangle)
+        {
+            // TODO: Not best algorithm
```
обычно при сдаче на ревью убирают все комментарии и тудушки

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+        private double startAngle = 0;
+        private readonly double radiusThreshold = 0;
+        private readonly double searchAngleStep = Math.PI / 2;
+        private readonly double offsetAngleStep = Math.PI / 180;
```
логика, касающаяся получения следующей точки на спирали не является ответственностью лэйаутера

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+            return rectangle;
+        }
+
+        public Point SearchBestOnAngle(double angle, double step, Size size)
```
название не очень понятное вышло

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+            return bestPoint;
+        }
+
+        private bool CanPlaced(Rectangle targetRectangle)
```
CanBePlaced

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+
+        private const double DoublePi = Math.PI * 2;
+        private double startAngle = 0;
+        private readonly double radiusThreshold = 0;
```
вот это все либо константы по своей сути, либо могли бы настраиваться как-то где-то

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+        public Point Center { get; }
+
+        private const double DoublePi = Math.PI * 2;
+        private double startAngle = 0;
```
тут можно не инициализировать 0, оно само так сделается, так как это дефолтное значение

goodoldlameme commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,97 @@
+                if (pointDistance < bestDistance)
+                {
+                    bestPoint = point;
+                    bestDistance = LinearMath.DistanceBetween(Center, bestPoint.CenterWith(size));
```
```LinearMath.DistanceBetween(Center, bestPoint.CenterWith(size))``` уже посчитана выше

goodoldlameme commented on cs/TagsCloudVisualization/Extensions/RectangleExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+                rectangle.Y + rectangle.Height / 2);
+        }
+
+        public static Point CenterWith(this Point point, Size size)
```
кажется не в тот класс попал метод

Folleach committed:
Structured code and create some stuff

- Create TagCloud for all cloud
	Such as: CircleTagCloud
- Create Render: FileCloudRender
- Create visualizers
	Suck as: SolidVisualizer, DistanceColorVisualizer

Folleach committed:
Remove magic TagsCloudVisualization.Interactive

Folleach committed:
Refactoring tests

Folleach committed:
Implement DistanceColorVisualizer and refactoring

Folleach committed:
RectangleVisualizer and fix CircularCloudLayouter

Folleach committed:
Create CircularPositioner

Folleach committed:
Add null check for size in CircularCloudLayouter

Folleach committed:
Add render test if test is fail

Also small refactoring of architecture:
- Add property TagCloud to IVisualizer

Alsooo add ImmutableTagCloud for drawing rectangles without layouters
and ImmutableViszualizer for it

Folleach committed:
Create README.md

Folleach committed:
Merge branch 'home-exercise' of https://github.com/Folleach/tdd into home-exercise

goodoldlameme commented on cs/TagsCloudVisualization/Visualizer/DistanceColorVisualizer.cs:
```diff
@@ -0,0 +1,57 @@
+using TagsCloudVisualization.TagClouds;
+
+namespace TagsCloudVisualization.Visualizer
+{
```
вообще, конечно, прикольно, что ты сделал несколько визуализаторов но на самом деле это часть следующей задачи.
тут такая тема: это только сейчас так повезло, что тебе это понадобится, но вообще делать то, что тебя не просят круто для себя, поковыряться, а в промышленной разработке такое не очень хорошо делать, так как это время, которое можно было бы потратить на что-то еще, а такая задача уже могла бы лежать в списке задач на будущее

goodoldlameme commented on cs/TagsCloudVisualization/Visualizer/DistanceColorVisualizer.cs:
```diff
@@ -0,0 +1,57 @@
+            }
+        }
+
+        private SolidBrush GetBrush(double fraction)
```
fraction не очень хорошее имя

Folleach committed:
Fix names and change IVisualizer to IVisualizer<T>

Folleach committed:
Removed dependency on layouter in TagCloud

Folleach committed:
Fix and add tests

Fixes:
- CircularCloudTests
Add:
- CircularPositionerTests
- Tests for rectangle extensions

Also add some extensions:
- Rectangle MaxDistanceFromCorner
- Rectangle GetCorners
